swtpc8212: fill in a lot more of this machine

68bit opened the pull request:
This patch gets it running. It adds the MCM66750 character generator, and implements the MC6845 row update function. Most of the I/O has been worked out with help from the CT-82 user manual. The various screen formats and the graphics modes appear to be working and plausible. Printer support has been added, and a beeper. A generic keyboard had been added, but the keyboard will need work. There appear to be keyboard cursor keys that are not implemented. Some of the keyboard control codes are not operational, can not be generated.

The ROMs for this one, with the added MCM66750 ROM, are at: https://68bit.org/swtpc8212.zip

I would like to be able to actually connect this to an emulated swtpc machine in mame and is that possible? If not then are there any plans to allow these rs232 devices to be connected? Might it be better as a rs232 slot device like the ie15 terminal? Is see some other terminals in mame, like the VT100, and it would be nice to be able to use these too, am I missing something?

cracyc commented:
>I would like to be able to actually connect this to an emulated swtpc machine in mame and is that possible?

You can attach a null modem to the serial port with "-rs232 null_modem" and a socket to the null modem with "-bitbngr socket.127.0.0.1:<portnum>".  If you run mame twice and both instances use the same port the first should listen and the second connect.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -1,19 +1,57 @@
+#include "speaker.h"
+#include "imagedev/printer.h"
+#include "machine/terminal.h"
+#include "bus/rs232/rs232.h"
```
Please keep the device includes above the ones from src/emu, and keep them sorted.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
 	void mem_map(address_map &map);
+	virtual void machine_reset() override;
+	virtual void machine_start() override;
+	virtual void device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr) override;
```
Please don't make overridden base members private - it makes it harder to make derived state classes for related machines.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
+	PORT_START("parity_select")
+	PORT_DIPNAME(0x80, 0x00, "Parity Select (Bit 7)")
+	PORT_DIPSETTING(0x00, "Odd or Mark (On)")
+	PORT_DIPSETTING(0x80, "Even or Space (Off)")
```
Please combine these into a single port, and use DIP locations to indicate their location rather than adding it to the name.

ajrhacker replied:
It should be noted that with terminals whose parity options are restricted to even, odd or none, the manuals often list the "no parity" switch separately.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
+	PORT_START("mode_switch")
+	PORT_DIPNAME(0x20, 0x00, "Mode switch")
+	PORT_DIPSETTING(0x00, "Conversational")
+	PORT_DIPSETTING(0x20, "Page edit")
```
Things that aren't DIP switches should use `CONF` macros instead.  Also, the jumpers should probably be a single port.

ajrhacker replied:
I usually use `PORT_DIPNAME` for jumpers because it allows locations to be specified. Jumper links and DIP switches are more or less identical electrically, and are occasionally substituted for each other.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
+	uint8_t parity_select = m_parity_select->read(); // PB7
+
+	return parity_select | no_parity | mode | baud_rate;
+}
```
This wouldn't be necessary if you combined these into a single port.

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -22,82 +60,359 @@ class swtpc8212_state : public driver_device
+
+	m_latch_data = 0x1f;
+
+	m_uart->cts_w(CLEAR_LINE);
```
This porbably shouldn't be here - CTS comes from the device attached to the port, and if that device is keeping CTS deasserted when the terminal is rese, the input won't magically change.

ajrhacker replied:
Most classic UARTs and USARTs automatically and unconditionally disable the transmitter when CTS is in a marking state (1). Many system designers disliked this behavior and tied the CTS pin to GND, with the actual RS232 CTS being read from some other port (e.g. VT100).

cuavas commented on src/mame/drivers/swtpc8212.cpp:
```diff
@@ -117,8 +448,10 @@ ROM_START( swtpc8212 ) // MC6802P, 2xMC6821P, INS8250N, MCM66750, MC6845P, bank
 ROM_END
 
-COMP(1980, swtpc8212, 0, 0, swtpc8212, swtpc8212, swtpc8212_state, empty_init, "Southwest Technical Products", "SWTPC 8212 Video Terminal", MACHINE_IS_SKELETON)
+COMP(1980, swtpc8212, 0, 0, swtpc8212, swtpc8212, swtpc8212_state, empty_init, "Southwest Technical Products", "SWTPC 8212 Video Terminal", MACHINE_NO_SOUND_HW )
```
A beeper is sound hardware - the flag isn't appropriate.

68bit commented:
Thanks for the review, and I'll try to address that.

I've tried an alternative abstraction, after studying the rs232 bus ie15 example, and now have the 8212 terminal implemented as a device with two front ends, one front end is a stand alone terminal and the other is on the rs232 bus. This appears to be much more convenient than trying to loop two separate processes via the rs232 loopback approach suggested above, and is going to be a lot clear running in a web browser.

Initial testing on the swtpc machine helped find the 'pia0' PA0 input is the 'duplex' of echo configuration option, and I'll add that too, it is needed to stop the echoing.

Not sure why the ie15 terminal used the diserial class, that seemed redundant, converting serial to parallel and back. So for the 8212 terminal the device just includes some public functions and callbacks to pass along the raw serial line state, and this seems simpler and works.

So would people be happy to add this 8212 terminal on the rs232 bus? Would it be welcomed if I put up an alternative patch implementing that to consider? Or do people think it would 'pollute' the rs232 bus to add such terminals, and others?

rb6502 commented:
No, I'd like to see all working terminals able to be RS232 devices.  In the future when a single MAME instance can run multiple machines, that will be important for connecting a terminal to an emulated host machine.  It just happens that the way things are now it's easier to connect to things outside of MAME if it's also available as a driver.

68bit commented:
Tried to address all the feedback, thanks. There have been a lot more changes, the bulk of the code has been moved to a device and there are now two 'front ends'. One connecting the terminal device to a rs232 port as before so it can be used stand alone. The other uses the device as a rs232 port bus slot so that it can be conveniently used as a terminal for other emulated PCs.

This 'two front ends' approach is taken from the exising ie15 terminal design, however the implementation is quite different. This proposal adds some public line functions and callbacks, as appears common for devices, and the front ends hook into these. This is in contrast to the ie15 design in which the device inherits the rs232 port interface and even a diserial interface - it just seems the wrong design, and it seems pointless to be converting from serial to parallel and back again.

There would appear to be potential to add a new class to define an interface for rs232 terminals, an interface that could be used to connect to a mame rs232 port on either side, the two front ends.

Still not happy with the searching for the ROMs, and need to explore a solution. It current searches for the ROMs using 'swtpc8212' for the stand alone front end, and 'swtpc8212-terminal' when used as a rs232 bus slot, an unnecessary duplication?

Feedback welcomed. I'll make another pass over it anyway.

smf- commented:
>This is in contrast to the ie15 design in which the device inherits the rs232 port interface and even a
> diserial interface - it just seems the wrong design, and it seems pointless to be converting from serial
> to parallel and back again.

What do you mean? Why do you think it's converting serial to parallel and back again?

68bit commented:
> > This is in contrast to the ie15 design in which the device inherits the rs232 port interface and even a
> > diserial interface - it just seems the wrong design, and it seems pointless to be converting from serial
> > to parallel and back again.
> 
> What do you mean? Why do you think it's converting serial to parallel and back again?

It uses the diserial interface which does that conversion, for example:

```
void ie15_terminal_device::rcv_complete()
{
	receive_register_extract();
	term_write(get_received_char());
}
```

Also note that the rs232 slot device has config options for the baud rate etc, these are for that conversion. This appears unnecessary, and in the proposed PR it simply passes along the serial data.

rb6502 commented:
> Still not happy with the searching for the ROMs, and need to explore a solution. It current searches for the ROMs using 'swtpc8212' for the stand alone front end, and 'swtpc8212-terminal' when used as a rs232 bus slot, an unnecessary duplication?
 
We've addressed this problem other places by making the guts of the device a standalone separate device and then both the driver and rs232 slot device are small shells that instantiate the main device and route the serial signals to either an rs232 port device (driver) or the bus slot (rs232 slot device).

I still don't get your objection to IE15.  Serial in MAME is internally represented as a real bitstream with the correct timings and start/stop/parity bits, because doing it that way solves some nasty problems.  diserial is used in IE15 because we don't (yet) support the UART chip it uses (plus several UART devices themselves use diserial internally).

68bit commented:
The 8212 terminal device could be separated further from the rs232 bus slot interface for it. There was no need to make that a sub class, and now the ROMs are consistently searched in swtpc8212_device.

cuavas commented on src/devices/machine/swtpc8212.cpp:
```diff
@@ -0,0 +1,475 @@
+
+#include "machine/input_merger.h"
+#include "machine/swtpc8212.h"
+#include "emu.h"
```
The PCH and module unit header should come first, otherwise there are going to be problems with different behaviour in different configurations.

cuavas commented on src/devices/machine/swtpc8212.cpp:
```diff
@@ -0,0 +1,475 @@
+READ8_MEMBER(swtpc8212_device::pia1_pb_r)
+{
+	return m_dip_switches->read();
+}
```
Why do you have this member here at all?  You can use `.set_ioport(...)` to connect a devcb directly to an I/O port, and avoid the trampoline.

cuavas commented on src/devices/machine/swtpc8212.cpp:
```diff
@@ -0,0 +1,475 @@
+	// CA2 - parallel printer, 'data ready' output.
+	// CB1 - Handshake input?
+	// CB2 - Handshake output?
+	pia6821_device &pia1(PIA6821(config, "pia1"));
```
When you have a device finder, you should use it in place of the literal tag and avoid the need for a temporary when building machine config.  It's less error-prone.  This applies to all the devices here that you have device finders for.

68bit commented:
Hopefully the feedback has been addressed, thanks. Also added a simply implementation of the keypad, filling in what is known. Did some testing, wrote a terminfo entry for this terminal and ran some of the ncurses tests etc and it seems ok for a start.

cuavas commented on src/devices/bus/rs232/swtpc8212.cpp:
```diff
@@ -0,0 +1,74 @@
+	output_rxd(1);
+	output_dcd(0);
+	output_dsr(0);
+	output_cts(0);
```
You shouldn't be doing this in `device_reset` – if the terminal actually changes line state on reset it should push the changes out itself.  Doing it in this shim device could cause the state seen by the other side of the port to get out of sync.

68bit replied:
This layer, the rs232 bus, is more than a straight-through connection of the terminal to the host, rather it actively directs and controls the lines, it is one of many forms of a null-modem cable, and have your considered it from that perspective?

Here, DCD and DSR are not being drive by the terminal, these need initializing, so these some  ok.

CTS is *optionally* driven from the terminal DTR line, but it might not be driven at all, it needs initializing, so this also seems ok.

Guess you might argue that initializing the rdx line is not necessary, but why not, it seems prudent.

If there is some state that could be lost here then can anyone point it out? It seems ok, in this instance, unlike some other goof's I've made! The bus interface 'output_*()' functions update the rs232 layer state variables, so they appear ok.

cuavas commented on src/devices/machine/swtpc8212.cpp:
```diff
@@ -0,0 +1,605 @@
+		program[0x00f0] = 0x0b; // Delete, was Crtl-Z, now Ctrl-K
+		program[0x00f1] = 0x00; // Scroll Down, was Crtl-O, now unused
+		program[0x00f2] = 0x0e; // Down, was Crtl-B, now Ctrl-N
+	}
```
I really don't like these kinds of patches.  It's not the MAME way.  Can you please do this using the cheat plugin, or some other way using Lua scripting?  (Yes, this is another reminder that we need a kind of "IPS for multiple ROM regions" format to support soft-patching via a plugin to allow ROM-only hacks to run in mainline MAME easily.)

68bit replied:
Yes, that had occurred to me. I am happy to pull that for now so people can consider it further if this is blocking this PR?

Here's my perspective. Systems in the late 1970's and early 1980's were limited, the ROMs and EPROMs were limited, there was a large DIY element to getting a computer running and fitting components together. Documentation commonly came with hardware mods to implement various functionality. Since the ROMs were limit, their functionality was limited, and it was not unexpected to have to patch a ROM here or there, it was not considered 'cheating', and even working with these systems today that is how it rolls.

If I had a 8212 terminal and wanted to connect it to a UniFlex SWTPC09 today then I would be patching the ROM to modify the codes send by the cursor keys - that's not a emulator hack, rather it's just how it rolls with this tech.

Here's another example. The CT-82 version 1 ROM supports 9600 baud, 1 stop bit, no parity, and that's it. Version 2 re-purposed the dips switches and jumpers to add a range of other formats, but all with 2 stop bits, there is no support for 1 stop bit and that breaks compatibility with many systems that this terminal was used with in the day. To get everything working together with this vintage tech does need some ROM patching, it's just how it rolled.

So given that, how to fit in with the 'MAME approach'?

1. Just use some representative ROM, potentially leaving the MAME emulators unworkable in practice. Would the MAME project even want required modifications noted in the source code or has that all become too much and are you wanting to separate the hardware emulation from the documentation of the software history and how to get things working together?

2. Multiple ROMs, for each combination of patches. This would appear to fit with the approach taken by some other machines in MAME. However where there are multiple patches there is an explosion of ROM combinations which is not great.

3. A run time loadable ROM, so people supply their own.

Approaches 2 and 3 leave the details of the patch to be explained elsewhere, rather than in the MAME emulator source code, and I was hoping to dump what knowledge I can in these emulators, I can't host a website anymore, people leave unfortunately.

4. ROM patches in the MAME emulator for some key trivial changes. This avoids having ROMs for each combination of patches. It gives some direction as to what changes have been made and where and why. This is the approach I have taken for these vintage systems.

68bit commented:
Removed the keypad ROM patch config option. Rebased, some trivial changes. Left the RS232 start initialization for now as it seems appropriate to me, but if someone has the patience to explain the issue then it will be addressed.

68bit commented:
Modified the RS232 initialization slightly, so only initialize the CTS line from the RS232 bus interface if there is no CTS connection, so when there is no CTS flow control. Also canonicalized the ioport names to uppercase.

cuavas commented on src/devices/bus/ss50/mps.cpp:
```diff
@@ -48,26 +50,34 @@ class ss50_mps_device : public device_t, public ss50_card_interface
+	PORT_START("CTS_ROUTE")
+	PORT_CONFNAME(1, 0, "CTS route")
+	PORT_CONFSETTING(0, "Wired low (standard)")
+	PORT_CONFSETTING(1, "Connected through (modified)")
```
This is another control line that can get stuck if the switch is flipped while running.  I won't hold up merging it any longer, but this one should be fixed. as well.
